File: leetcode-practice-typescript/src/array/238-product-of-array-except-self.ts
function productExceptSelf(nums: number[]): number[] {
    // 先用answer表示下标i左侧的所有数的乘积
    const n = nums.length;
    const answer = Array(n);
    answer[0] = 1;
    for (let i = 1; i < n; i++) {
        answer[i] = answer[i - 1] * nums[i - 1];
    }
    // 再逐个更新右侧的乘积
    let rightP = 1;
    for (let i = n - 2; i >= 0; i--) {
        rightP *= nums[i + 1];
        answer[i] *= rightP;
    }
    return answer;
}

console.log(productExceptSelf([1, 2, 3, 4]));

File: leetcode-practice-python/src/array/238_product_of_array_except_self.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        n = len(nums)
        answer = [1] * n

        # 第一次遍历，计算不包含下标i的前缀积
        for i in range(1, n):
            answer[i] = answer[i - 1] * nums[i - 1]

        # 第二次遍历，计算后缀积，并乘以前缀积得到结果
        suffix = 1
        for i in range(n - 1, -1, -1):
            answer[i] = answer[i] * suffix
            suffix = suffix * nums[i]

        return answer


if __name__ == '__main__':
    s = Solution()
    print(s.productExceptSelf([1, 2, 3, 4]))
